### ISSUE TITLE
RFC: canonicalize random access as a global compiler pass

### ISSUE THREAD

**Issue by willow-ahrens** [willow-ahrens/Finch.jl#223]
I'm posting this as an RFC because I'd like feedback on whether this might be a blocker for anyone interested in scheduling, mainly @wraith1995

In essence, I'd like to run a transformation pass at the beginning of the program which transforms expressions like
```julia
for i = 1:10
   B[] += A[I[i]]
end
```
into expressions like:
```julia
for i = 1:10
   s = I[i]
   for j = s:s
      B[] += A[j]
   end
end
```

This is roughly what we do already, but we do this lazily (only when `I[i]` becomes evaluated will we request `selectstyle` for it and perform this transformation just in time).

I'd like to make the change to the explicit global transformation because it allows me to enforce a canonical form where the only type that can go into an array access expression is an index.

Is this change acceptable to everyone?

**Comment by wraith1995** [comment]
This is fine with me.

**Comment by wraith1995** [comment]
@willow-ahrens An additional thought on this: to support things like Taichi's index computations or things like Jess' stuff, this is a good change as it could allow us to later try to analyze groups of accesses e.g. `A[I[i]] +  A[I[i]+1]` or `A[I[i]] +  A[I[i] + I[i] % 5]`

**Comment by willow-ahrens** [comment]
Fixed in #231
